### PR TITLE
[Finished] TypeScript conversion

### PR DESCRIPTION
**Pull request by lukastaegert**
*Everything is typed now and this is ready to be merged!*

Closes #1570.

This is based off @Rich-Harris original PR #1570. @guybedford and myself have been working the last days to add missing types and make everything as smooth as possible so that we can do the transformation as soon as possible.

Why TypeScript, why now?

Cons:
* Rollup is a JavaScript bundler that used to be able to bundle itself (nearly) without additional tools
* Contributors already know the language

Pros:
* TypeScript is not very difficult if you know JavaScript - basically everything you could do before is still possible using the same syntax, just some added types here and there. And with the added TypeScript plugin, Rollup can still bundle itself.
* We NEED profound refactorings and architectural improvements, especially around the `Bundle.js`/`Module.js` complex. The internal API surface of those is HUGE and only understood by few - if at all. TypeScript allows us to make members private and define interfaces to structure the API around
* We also want to introduce new abstraction layers that might in the end even make code splitting possible. Again, TypeScript is helping a lot here

### CODE REVIEW

**Comment by guybedford** [comment]
Typing PRs:
* Magic String: https://github.com/Rich-Harris/magic-string/pull/131
* Plugin TypeScript for ES5 Iteration support: https://github.com/rollup/rollup-plugin-typescript/pull/102

**Comment by ankeetmaini** [comment]
This looks amazing! I'd be lot more confident in changing the code now. Let me know if you need a ✋, I'm not a TS expert though!

**Comment by JamesHenry** [comment]
I am little biased, but I would strongly recommend taking this opportunity to also introduce Prettier into your toolchain, and never think about formatting ever again :) Let me know if you have any questions about it

**Comment by eventualbuddha** [comment]
Whoa! I didn't know about this, but I'm a big fan of TypeScript so definitely 💯 from me.

File: src/ast/nodes/CallExpression.ts [hunk 1]
```diff
@@ -5,6 +5,7 @@ import ExecutionPathOptions from '../ExecutionPathOptions';
 import SpreadElement from './SpreadElement';
 import { PredicateFunction } from '../values';
 import GlobalVariable from '../variables/GlobalVariable';
+import { ObjectPath } from '../variables/VariableReassignmentTracker';
 
 // TODO: 3 typing failures because AwaitExpression has no forEachReturnExpressionWhenCalledAtPath
```
Comment (guybedford) [inline src/ast/nodes/CallExpression.ts]: Is this case resolved now? If so the comment here can be removed.

**Comment by guybedford** [comment]
I've just noticed that despite the "es5" target we've got in the TypeScript config, Rollup still assumes things like Map and iterator support.

Node 4 will support this and the const stuff so I was wondering if it's worth just updating the build to ES6 at this point? We can still individually pick used features based on what we deem important?

This came up in the process of trying to add downlevelIteration support for true ES5, and noticing that no Map shims were in place.

**Comment by guybedford** [comment]
This also reduces the TypeScript browser build when minified by 5KB.

File: src/ast/nodes/CallExpression.ts (hunk 1 above)
Comment (guybedford) [reply src/ast/nodes/CallExpression.ts]: I've removed this now.

**Comment by guybedford** [comment]
Ok I've reverted this attempt for now.

The Buble register made this seem like it could work, but it seems that es5 can only be dropped when dropping support for Node4 I guess.

Iteration support through https://github.com/rollup/rollup-plugin-typescript/pull/102 is no longer necessary though.

**Comment by lukastaegert** [comment]
Ok, this goes into master now...

**Comment by lukastaegert** [comment]
So, that's it. The next release will be a TypeScript release!

**Comment by dherges** [comment]
How cool is that! 🎇

**Comment by danbucholtz** [comment]
I just wanted to chime in and say you folks rock 🤘 .

Thanks,
Dan

**Pull request merged**